### PR TITLE
Neighbor Lists Updates and Dynamic Kernel bug fixes

### LINKED ISSUE #18: 🚀[FEA]: Let batch_naive_neighbor_list handle unwrapped coodinates just like the batch_cell_list
### Is this a new feature, an improvement, or a change to existing functionality?

Improvement

### How would you describe the priority of this feature request

Medium

### Please provide a clear description of problem you would like to solve.

It seems the `batch_naive_neighbor_list` gives incorrect graph if unwrapped coordinates are used. This results in models give wrong values as reported [here](https://github.com/TorchSim/torch-sim/issues/423). Currently, I have [patched](https://github.com/TorchSim/torch-sim/pull/436) models to always pass wrapped coordinates but it would be nice to have same behavior as the cell list implementation.

I think this is documented behavior in the code but since models compute edge vectors etc internally sometimes they also need a way to have access to the correct wrapped coordinates while the simulation maybe in unwrapped coordinates.

```
import torch
from ase.io import read as ase_read
from nvalchemiops.neighborlist import batch_cell_list, batch_naive_neighbor_list


def compute_distances(positions, cell, mapping, shifts_idx):
    shift_vectors = torch.einsum("ij,kj->ik", shifts_idx.to(cell.dtype), cell)
    displacements = positions[mapping[1]] - positions[mapping[0]] + shift_vectors
    return torch.norm(displacements, dim=1)


def compare(name, nl_fn, pos_w, pos_u, cell, pbc, batch_idx, cutoff):
    # Wrapped
    res_w = nl_fn(
        positions=pos_w,
        cutoff=cutoff,
        batch_idx=batch_idx,
        cell=cell,
        pbc=pbc,
        return_neighbor_list=True,
    )
    map_w = res_w[0]
    shift_w = res_w[2] if len(res_w) == 3 else torch.zeros(map_w.shape[1], 3)

    # Unwrapped
    res_u = nl_fn(
        positions=pos_u,
        cutoff=cutoff,
        batch_idx=batch_idx,
        cell=cell,
        pbc=pbc,
        return_neighbor_list=True,
    )
    map_u = res_u[0]
    shift_u = res_u[2] if len(res_u) == 3 else torch.zeros(map_u.shape[1], 3)

    dist_w = compute_distances(pos_w, cell[0], map_w, shift_w)
    dist_u = compute_distances(pos_u, cell[0], map_u, shift_u)

    n_match = map_w.shape[1] == map_u.shape[1]
    dist_match = n_match and torch.allclose(
        torch.sort(dist_w)[0], torch.sort(dist_u)[0], atol=1e-5
    )

    print(f"{name}: (wrapped={map_w.shape[1]}, unwrapped={map_u.shape[1]}) {dist_match}")


def main():
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    dtype = torch.float64
    cutoff = 5.0

    # Load structures
    atoms_w = ase_read("POSCAR_wrapped.vasp", format="vasp")
    atoms_u = ase_read("POSCAR_unwrapped.vasp", format="vasp")

    # Prepare tensors - ensure contiguous
    pos_w = torch.tensor(atoms_w.positions, device=device, dtype=dtype).contiguous()
    pos_u = torch.tensor(atoms_u.positions, device=device, dtype=dtype).contiguous()
    cell = (
        torch.tensor(atoms_w.cell.array.T, device=device, dtype=dtype)
        .unsqueeze(0)
        .contiguous()
    )
    pbc = (
        torch.tensor(atoms_w.pbc, device=device, dtype=torch.bool)
        .unsqueeze(0)
        .contiguous()
    )
    batch_idx = torch.zeros(len(atoms_w), device=device, dtype=torch.int32)

    print(f"Atoms: {len(atoms_w)}, Device: {device}, Cutoff: {cutoff} Å")
    compare(
        "batch_naive_neighbor_list",
        batch_naive_neighbor_list,
        pos_w,
        pos_u,
        cell,
        pbc,
        batch_idx,
        cutoff,
    )
    compare(
        "batch_cell_list", batch_cell_list, pos_w, pos_u, cell, pbc, batch_idx, cutoff
    )


if __name__ == "__main__":
    main()```
```

would give:

```
Atoms: 120, Device: cpu, Cutoff: 5.0 Å
batch_naive_neighbor_list: (wrapped=4032, unwrapped=2170) False
batch_cell_list: (wrapped=4032, unwrapped=4032) True
```

[POSCAR_wrapped.txt](https://github.com/user-attachments/files/25040050/POSCAR_wrapped.txt)

[POSCAR_unwrapped.txt](https://github.com/user-attachments/files/25040051/POSCAR_unwrapped.txt)

### Describe any alternatives you have considered

_No response_

### PR DESCRIPTION
**Pull request by dallasfoster**
# ALCHEMI Toolkit-Ops Pull Request

## Type of Change

- [ x ] Bug fix (non-breaking change that fixes an issue)
- [ x ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Performance improvement
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Related Issues

Fixes #18 

## Changes Made

- Fixes scalar types in NPT and NPH kernels to tensor types
- Allows for rebuild_flag in (batch) neighborlists and adds batch rebuild check.
- Fixes positions wrapping for naive neighbor list.

## Testing

- [ x ] Unit tests pass locally (`make pytest`)
- [ x ] Linting passes (`make lint`)
- [ x ] New tests added for new functionality meets coverage expectations?

## Checklist

- [ x ] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [ ] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [ x ] I have performed a self-review of my code
- [ x ] I have added docstrings to new functions/classes
- [ ] I have updated the documentation (if applicable)

## Additional Notes

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

File: nvalchemiops/dynamics/integrators/nose_hoover.py [hunk 1]
```diff
@@ -205,9 +205,9 @@ def _batch_compute_2ke_kernel(
 
 @wp.kernel
 def _nhc_compute_masses_kernel(
```
Comment (laserkelvin) [inline nvalchemiops/dynamics/integrators/nose_hoover.py:207]: Is there an advantage of having an unbatched version of this kernel?

File: nvalchemiops/dynamics/utils/tile_reductions.py [hunk 2]
```diff
@@ -0,0 +1,646 @@
...
+standard atomic operations.
+
+**Key Features:**
+- Shared memory indexed accumulation (reduces global atomic contention)
+- Indexed tile sum for batched per-system reductions
+- Warp-level shuffle reductions (no shared memory needed)
+- Full gradient support via adjoint snippets for automatic differentiation
+- Float32 and float64 variants for all utilities
+
+**Expected Performance Improvements:**
```
Comment (laserkelvin) [inline nvalchemiops/dynamics/utils/tile_reductions.py]: Remove this from the docstring

File: nvalchemiops/jax/neighbors/batch_naive.py [hunk 3]
```diff
@@ -306,20 +396,64 @@ def batch_naive_neighbor_list(
                     "Please provide max_atoms_per_system explicitly when using jax.jit."
                 ) from None
 
-        neighbor_matrix, neighbor_matrix_shifts, num_neighbors = _jax_fill_pbc(
+        # Pre-wrap positions: compute inv_cell then wrap
+        inv_cell = jnp.zeros_like(cell)
+        (inv_cell,) = _jax_inv_cells(
```
Comment (laserkelvin) [inline nvalchemiops/jax/neighbors/batch_naive.py]: This feels awfully complicated to call the jax bindings of the warp kernel to compute the inverse. Should we just have this call `jnp.linalg.inv` instead?

File: nvalchemiops/neighbors/batch_cell_list.py [hunk 4]
```diff
@@ -430,59 +431,8 @@ def _batch_cell_list_build_neighbor_matrix(
...
-    - Each atom is only paired with atoms from its own system
-    """
-    atom_idx = wp.tid()
-
-    # Find which system this atom belongs to
-    system_idx = batch_idx[atom_idx]
-
-    # Get system and atom specific parameters
+    half_fill: wp.bool,
+):
```
Comment (laserkelvin) [inline nvalchemiops/neighbors/batch_cell_list.py:435]: Need some short docstring?

File: nvalchemiops/neighbors/batch_naive.py [hunk 5]
```diff
@@ -623,31 +643,67 @@ def batch_naive_neighbor_matrix_pbc(
...
+    wp_vec_dtype = (
+        wp.vec3f
+        if wp_dtype == wp.float32
+        else wp.vec3d
+        if wp_dtype == wp.float64
+        else wp.vec3h
+        if wp_dtype == wp.float16
+        else None
+    )
+    inv_cell = wp.empty((cell.shape[0],), dtype=wp_mat_dtype, device=device)
```
Comment (laserkelvin) [inline nvalchemiops/neighbors/batch_naive.py]: Do `empty_like`?

**Comment by dallasfoster** [comment]
/ok to test [23f6aff](https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/38/commits/23f6afff79837d91ded694a2d8a5b82ee4d2b808)

File: nvalchemiops/neighbors/batch_cell_list.py (hunk 4 above)
Comment (dallasfoster) [reply nvalchemiops/neighbors/batch_cell_list.py:435]: added

File: nvalchemiops/jax/neighbors/batch_naive.py (hunk 3 above)
Comment (dallasfoster) [reply nvalchemiops/jax/neighbors/batch_naive.py]: done.
